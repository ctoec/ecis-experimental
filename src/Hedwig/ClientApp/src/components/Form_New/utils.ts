import moment from 'moment';

export function parseDateChange(event: React.ChangeEvent<any>) {
	return event.target.value ? moment.utc(parseInt(event.target.value, 10)).toDate() : null;
}
