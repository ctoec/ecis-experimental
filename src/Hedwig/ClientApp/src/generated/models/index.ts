export * from './Age';
export * from './CdcReport';
export * from './Child';
export * from './Enrollment';
export * from './Family';
export * from './FamilyDetermination';
export * from './Funding';
export * from './FundingSource';
export * from './FundingSpace';
export * from './FundingTime';
export * from './Gender';
export * from './Organization';
export * from './OrganizationPermission';
export * from './OrganizationReport';
export * from './ProblemDetails';
export * from './Region';
export * from './ReportingPeriod';
export * from './Site';
export * from './SitePermission';
export * from './User';
export * from './ValidationError';
export * from './ValidationProblemDetails';
