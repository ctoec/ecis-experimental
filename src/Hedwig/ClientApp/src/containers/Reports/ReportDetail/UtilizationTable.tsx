import React from 'react';
import Table, { TableProps } from '../../../components/Table/Table';
import { CdcReport, Age, FundingTime, Region } from '../../../generated';
import idx from 'idx';
import moment from 'moment';
import { CdcRates } from './CdcRates';
import { prettyAge } from '../../../utils/ageGroupUtils';
import { prettyFundingTime } from '../../../utils/fundingTimeUtils';
import currencyFormatter from '../../../utils/currencyFormatter';

export function calculateRate(accredited: boolean, titleI: boolean, region: Region, ageGroup: Age, time: FundingTime) {
  const rate = CdcRates.find(rate =>
    rate.accredited === accredited &&
    rate.titleI === titleI &&
    rate.region === region &&
    rate.ageGroup === ageGroup &&
    rate.time === time
  );

  return rate ? rate.rate : 0;
}

interface UtilizationTableRow {
  key: string
  ageGroup?: Age
  fundingTime?: FundingTime
  count: number
  capacity: number
  rate?: number
  total: number
  balance: number
}

export default function UtilizationTable(report: CdcReport) {
  const periodStart = idx(report, _ => _.reportingPeriod.periodStart);
  const periodEnd = idx(report, _ => _.reportingPeriod.periodEnd);
  const weeksInPeriod = periodStart && periodEnd ? moment(periodEnd).add(1, 'day').diff(periodStart, 'weeks') : 0;

  const accredited = report.accredited;
  const fundingSpaces = idx(report, _ => _.organization.fundingSpaces) || [];

  const sites = idx(report, _ => _.organization.sites) || [];
  const ageGroups = [Age.InfantToddler, Age.Preschool, Age.SchoolAge];
  const fundingTimes = [FundingTime.Full, FundingTime.Part];

  const rows: UtilizationTableRow[] = sites.flatMap(site => {
    const region = site.region;
    const titleI = site.titleI;

    const enrollments = site.enrollments.flatMap(enrollment => {
      const cdcFunding = enrollment.fundings.find(funding => funding.source === report.type);
      if (!cdcFunding) { return []; }
      return {
        ageGroup: enrollment.ageGroup,
        time: cdcFunding.time,
      };
    });

    return ageGroups.flatMap(ageGroup => fundingTimes.flatMap(fundingTime => {
      const count = enrollments
        .filter(enrollment => enrollment.ageGroup === ageGroup && enrollment.time === fundingTime)
        .length;

      const capacity = idx(
        fundingSpaces.find(space => space.ageGroup === ageGroup && space.time === fundingTime),
        _ => _.capacity
      ) || 0;

      const cappedCount = Math.min(count, capacity);

      const rate = calculateRate(accredited, titleI, region, ageGroup, fundingTime);

      const total = cappedCount * rate * weeksInPeriod;
      const paid = capacity * rate * weeksInPeriod;
      const balance = total - paid;

      return {
        key: `${ageGroup}-${fundingTime}`,
        ageGroup,
        fundingTime,
        count,
        capacity,
        rate,
        total,
        balance,
      }
    }));
  });

  const totalRow = rows.reduce((total, row) => {
    return {
      ...total,
      count: total.count + row.count,
      capacity: total.capacity + row.capacity,
      total: total.total + row.total,
      balance: total.balance + row.balance,
    };
  }, { key: 'total', count: 0, capacity: 0, total: 0, balance: 0 });

  rows.push(totalRow);

  const tableProps: TableProps<UtilizationTableRow> = {
    id: 'utilization-table',
    data: rows.filter(row => row.count || row.capacity),
    rowKey: row => row.key,
    columns: [
      {
        name: '',
        cell: ({ row }) => (
          <th>
            <strong>
              {row.key === 'total' ? 'Total' : `${prettyAge(row.ageGroup)}`}
            </strong>
            {row.fundingTime && (
              <> &ndash; {prettyFundingTime(row.fundingTime)}</>
            )}
          </th>
        )
      },
      {
        name: 'Utilization',
        cell: ({ row }) => (
          <td className={
            (row.key === "total" ? "oec-table__cell--strong" : "") + " " +
            (row.count !== row.capacity ? "oec-table__cell--red" : "")
          }>
            {row.count}/{row.capacity} spaces
          </td>
        )
      },
      {
        name: <>Reimbursement <br />rate (weekly)</>,
        cell: ({ row }) => (
          <td>
            {row.key !== 'total' && currencyFormatter(row.rate || 0)}
          </td>
        )
      },
      {
        name: `Total (${weeksInPeriod} weeks)`,
        cell: ({ row }) => (
          <td className={row.key === "total" ? "oec-table__cell--strong" : ""}>
            {currencyFormatter(row.total)}
          </td>
        )
      },
      {
        name: `Balance`,
        cell: ({ row }) => (
          <td className={
            (row.key === "total" ? "oec-table__cell--strong" : "") + " " +
            (row.balance < 0 ? "oec-table__cell--red" : "")
          }>
            {row.balance < 0 ? '(' : ''}
            {currencyFormatter(Math.abs(row.balance))}
            {row.balance < 0 ? ')' : ''}
          </td>
        )
      }
    ]
  }

  return <>
    <h2 className="margin-bottom-neg-205">Monthly utilization</h2>
    <Table {...tableProps} fullWidth />
  </>
}
