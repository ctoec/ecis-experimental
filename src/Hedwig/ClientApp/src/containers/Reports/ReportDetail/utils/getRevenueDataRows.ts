import { CdcReport, FundingTime, Enrollment, FundingSpace, FundingSource, Age } from "../../../../generated";
import idx from "idx";
import { getReportingPeriodWeeks, getFundingSpaces, fundingSpaceSorter } from "../../../../utils/models";
import { getSplitUtilizations } from "../../../../utils/models/fundingTimeSplitUtilization";
import { DeepNonUndefineableArray } from "../../../../utils/types";
import { countFundedEnrollments, calculateRate, productOfUnknowns } from "../../../../utils/utilizationTable";

export interface RevenueDataRow  {
	key: string;
	ageGroup?: Age;
	fundingTime?: FundingTime;
	count: number;
	capacity: number;
	ftRate?: number;
	ptRate?: number;
	total: number;
	balance: number;
	weeksSplit?: {
		partWeeks: number | undefined;
		fullWeeks: number | undefined;
	};
}
export const getRevenueDataRows = (report: CdcReport) => {
	// TODO: if the space lives on the organization but the rates live on the site,
	// we need to reconsider how we handle multi site org rate calculation
	const site = idx(report, (_) => _.organization.sites[0]);
	const fundingSpaces = idx(report, (_) => _.organization.fundingSpaces);

	if (!site || !fundingSpaces) return [];

	const weeksInPeriod = getReportingPeriodWeeks(report.reportingPeriod);
	const splitTimeFundingSpaces = getFundingSpaces(
		report.organization && report.organization.fundingSpaces,
		{
			time: FundingTime.Split,
		}
	);
	const timeSplitUtilizations = getSplitUtilizations(report, splitTimeFundingSpaces, weeksInPeriod);

	const enrollments = (idx(report, (_) => _.enrollments) || []) as Enrollment[];

	const cdcFundingSpaces = (fundingSpaces as DeepNonUndefineableArray<FundingSpace>)
		.sort(fundingSpaceSorter)
		.filter((fundingSpace) => fundingSpace.source === FundingSource.CDC);

	let rows: RevenueDataRow[] = cdcFundingSpaces.map((space) => {
		const capacity = space.capacity;
		const ageGroup = space.ageGroup;
		const fundingTime = space.time;
		const count = countFundedEnrollments(enrollments, space.id);
		const ptRate = calculateRate(
			report.accredited,
			site.titleI,
			site.region,
			ageGroup,
			FundingTime.Part
		);
		const ftRate = calculateRate(
			report.accredited,
			site.titleI,
			site.region,
			ageGroup,
			FundingTime.Full
		);

		let fullWeeks: number | undefined, partWeeks: number | undefined;
		if (fundingTime === FundingTime.Split) {
			const thisUtilization = (timeSplitUtilizations || []).find((u) => u.reportId === report.id);
			fullWeeks = 0;
			partWeeks = 0;
			if (thisUtilization) {
				fullWeeks = thisUtilization.fullTimeWeeksUsed || 0;
				partWeeks = thisUtilization.partTimeWeeksUsed || 0;
			}
		} else if (fundingTime === FundingTime.Full) {
			fullWeeks = weeksInPeriod;
		} else if (fundingTime === FundingTime.Part) {
			partWeeks = weeksInPeriod;
		}

		const cappedCount = Math.min(count, capacity);
		const total =
			productOfUnknowns([cappedCount, fullWeeks, ftRate]) +
			productOfUnknowns([cappedCount, partWeeks, ptRate]);
		const paid =
			productOfUnknowns([capacity, fullWeeks, ftRate]) +
			productOfUnknowns([capacity, partWeeks, ptRate]);

		const balance = total - paid;

		return {
			key: `${ageGroup}-${fundingTime}`,
			ageGroup,
			fundingTime,
			count,
			capacity,
			ftRate,
			ptRate,
			total,
			balance,
			weeksSplit: {
				partWeeks: partWeeks,
				fullWeeks: fullWeeks,
			},
		};
	});

	const totalRow = rows.reduce(
		(total, row) => {
			return {
				...total,
				count: total.count + row.count,
				capacity: total.capacity + row.capacity,
				total: total.total + row.total,
				balance: total.balance + row.balance,
			};
		},
		{
			key: 'total',
			count: 0,
			capacity: 0,
			total: 0,
			balance: 0,
		}
	);

	rows.push(totalRow);

	return rows;
};
