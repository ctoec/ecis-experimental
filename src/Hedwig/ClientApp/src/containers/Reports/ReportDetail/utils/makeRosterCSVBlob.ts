import { parse as parseCsv } from 'json2csv';
import { Enrollment } from "../../../../generated";
import { nameFormatter } from "../../../../utils/stringFormatters";
import dateFormatter from "../../../../utils/dateFormatter";
import { prettyMultiRace, prettyEthnicity, prettyGender, prettyAge, prettyFundingSpaceTime, reportingPeriodFormatter } from "../../../../utils/models";
import { propertyDateSorter } from "../../../../utils/dateSorter";

export const  makeRosterCSVBlob = (enrollments: Enrollment[] | undefined | null) => {
	enrollments = enrollments || [];

	const fields = [
		{
			label: 'Name',
			value: (row: Enrollment) => nameFormatter(row.child),
		},
		{
			label: 'Birthdate',
			value: (row: Enrollment) => dateFormatter(row.child?.birthdate),
		},
		{
			label: 'Race',
			value: (row: Enrollment) => prettyMultiRace(row.child, { showAll: true }),
		},
		{
			label: 'Ethnicity',
			value: (row: Enrollment) => prettyEthnicity(row.child),
		},
		{
			label: 'Gender',
			value: (row: Enrollment) => prettyGender(row.child?.gender),
		},
		{
			label: 'Lives with foster family',
			value: (row: Enrollment) => (row.child?.foster ? 'Yes' : 'No'),
		},
		{
			label: 'Experienced homelessness or housing insecurity',
			value: (row: Enrollment) => (row.child?.family?.homelessness ? 'Yes' : 'No'),
		},
		{
			label: 'Household size',
			value: (row: Enrollment) =>
				row.child?.foster
					? 'Exempt'
					: row.child?.family?.determinations?.length
					? row.child?.family?.determinations[0].numberOfPeople
					: 'No data',
		},
		{
			label: 'Annual household income',
			value: (row: Enrollment) =>
				row.child?.foster
					? 'Exempt'
					: row.child?.family?.determinations?.length
					? row.child?.family?.determinations[0].income
					: 'No data',
		},
		{
			label: 'Determination date',
			value: (row: Enrollment) =>
				row.child?.foster
					? 'Exempt'
					: row.child?.family?.determinations?.length
					? dateFormatter(row.child?.family?.determinations[0].determinationDate)
					: 'No data',
		},
		{
			label: 'Enrollment start date',
			value: (row: Enrollment) => dateFormatter(row.entry),
		},
		{
			label: 'Site',
			value: (row: Enrollment) => row.site?.name || '',
		},
		{
			label: 'Age group',
			value: (row: Enrollment) => prettyAge(row.ageGroup),
		},
		{
			label: 'Funding',
			value: (row: Enrollment) => {
				const [mostRecentFunding] = (row.fundings || []).sort((a, b) =>
					propertyDateSorter(a, b, (f) => f?.firstReportingPeriod?.period)
				);
				return mostRecentFunding
					? `${prettyAge(row.ageGroup)} - ${prettyFundingSpaceTime(mostRecentFunding.fundingSpace)}`
					: 'No data';
			},
		},
		{
			label: 'First reporting period',
			value: (row: Enrollment) => {
				const [mostRecentFunding] = row.fundings || [];
				return mostRecentFunding
					? reportingPeriodFormatter(mostRecentFunding.firstReportingPeriod)
					: 'No data';
			},
		},
		{
			label: 'Last reporting period',
			value: (row: Enrollment) => {
				const [mostRecentFunding] = row.fundings || [];
				return mostRecentFunding
					? mostRecentFunding.lastReportingPeriod
						? reportingPeriodFormatter(mostRecentFunding.lastReportingPeriod)
						: 'ongoing'
					: 'No data';
			},
		},
		{
			label: 'Recieves Care 4 Kids',
			value: (row: Enrollment) => {
				const [mostRecentCert] = (row.child?.c4KCertificates || []).sort((a, b) =>
					propertyDateSorter(a, b, (c) => c.startDate)
				);
				return mostRecentCert && (!mostRecentCert.endDate || mostRecentCert.endDate > new Date())
					? 'Yes'
					: 'No';
			},
		},
		{
			label: 'Care 4 Kids ID',
			value: 'child.c4KFamilyCaseNumber',
			default: '',
		},
		{
			label: 'Care 4 Kids certificate start date',
			value: (row: Enrollment) => {
				const [mostRecentCert] = row.child?.c4KCertificates || [];
				return mostRecentCert
					? mostRecentCert.startDate
						? dateFormatter(mostRecentCert.startDate)
						: 'No data'
					: '';
			},
		},
	];

	const enrollmentsCsv = parseCsv(enrollments, { fields });
	return new Blob([enrollmentsCsv], { type: 'text/csv' });
};
