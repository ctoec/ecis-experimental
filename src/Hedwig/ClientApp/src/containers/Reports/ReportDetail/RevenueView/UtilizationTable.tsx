import React from 'react';
import { Table, TableProps, useGenericContext, FormContext } from '@ctoec/component-library';
import {
	CdcReport,
} from '../../../../generated';
import {
	prettyAge,
	prettyFundingTime,
	getReportingPeriodWeeks,
} from '../../../../utils/models';
import currencyFormatter from '../../../../utils/currencyFormatter';
import cx from 'classnames';
import {
	makePrefixerFunc,
	ReimbursementRateLine,
} from '../../../../utils/utilizationTable';
import { RevenueDataRow, getRevenueDataRows } from '../utils/getRevenueDataRows';

type UtilizationTableRow = RevenueDataRow & {
	showFundingTime?: boolean;
	// Used for numeral alignment formatting
	maxes: {
		total: number;
		rate: number;
		balance: number;
	};
}

export default function UtilizationTable() {
	const { data: report } = useGenericContext<CdcReport>(FormContext);

	// Transform RevenueDataRow into UtilizationTableRow
	let rows: UtilizationTableRow[] = getRevenueDataRows(report).map((row) => ({
		...row,
		showFundingTime: (report.organization?.fundingSpaces || []).filter((fs) => fs.ageGroup === row.ageGroup).length > 1,
		maxes: {
			total: row.total,
			balance: row.balance,
			rate: Math.max(row.ptRate || 0, row.ftRate || 0)
		}
	}));

	// And do reduce transformation for total row
	let totalRow = rows[rows.length - 1];
	totalRow = rows.reduce(
		(total, row) => ({
			...total,
			maxes: {
				total: Math.max(total.total, row.total),
				rate: Math.max(total.maxes.rate, row.ftRate || 0, row.ptRate || 0),
				balance: Math.max(Math.abs(total.balance), Math.abs(row.balance))
			}
		}),
		totalRow
	)

	// Get prefixers
	const reimbursementPrefixer = makePrefixerFunc(totalRow.maxes.rate);
	const totalPrefixer = makePrefixerFunc(totalRow.maxes.total);
	const balancePrefixer = makePrefixerFunc(totalRow.maxes.balance);

	// Instantiate table props
	const tableProps: TableProps<UtilizationTableRow> = {
		id: 'utilization-table',
		data: rows,
		rowKey: (row) => row.key,
		columns: [
			{
				name: '',
				cell: ({ row }) => (
					<th>
						<strong>{row.key === 'total' ? 'Total' : `${prettyAge(row.ageGroup)}`}</strong>
						{row.fundingTime && row.showFundingTime && (
							<> &ndash; {prettyFundingTime(row.fundingTime, { splitTimeText: 'pt/ft split' })}</>
						)}
					</th>
				),
			},
			{
				name: 'Utilization',
				cell: ({ row }) => (
					<td
						className={
							(row.key === 'total' ? 'oec-table__cell--strong' : '') +
							' ' +
							(row.count !== row.capacity ? 'oec-table__cell--red' : '')
						}
					>
						{row.count}/{row.capacity} spaces
					</td>
				),
			},
			{
				name: 'Reimbursement rate',
				className: 'text-right',
				cell: ({ row }) => {
					const { partWeeks, fullWeeks } = row.weeksSplit || {};
					return (
						<td className="text-tabular text-right">
							<ReimbursementRateLine
								prefix={reimbursementPrefixer(row.ptRate || 0)}
								prettyRate={currencyFormatter(row.ptRate || 0, true)}
								weeksInPeriod={partWeeks}
								suffix="(pt)"
							/>
							<ReimbursementRateLine
								prefix={reimbursementPrefixer(row.ftRate || 0)}
								prettyRate={currencyFormatter(row.ftRate || 0, true)}
								weeksInPeriod={fullWeeks}
								suffix="(ft)"
							/>
						</td>
					);
				},
			},
			{
				name: `Total (${getReportingPeriodWeeks(report.reportingPeriod)} weeks)`,
				className: 'text-right',
				cell: ({ row }) => {
					const prefix = totalPrefixer(row.total);
					return (
						<td
							className={cx(
								{ 'oec-table__cell--strong': row.key === 'total' },
								'text-tabular',
								'text-right'
							)}
						>
							<span>$ </span>
							<span style={{ visibility: 'hidden' }}>{prefix}</span>
							{currencyFormatter(row.total, true)}
						</td>
					);
				},
			},
			{
				name: 'Balance',
				className: 'text-right',
				cell: ({ row }) => {
					const prefix = balancePrefixer(row.balance);
					return (
						<td
							className={cx(
								{ 'oec-table__cell--strong': row.key === 'total' },
								{ 'oec-table__cell--red': row.balance < 0 },
								'text-tabular',
								'text-right'
							)}
						>
							<div className={cx('position-relative', { 'one-half-char-right': row.balance < 0 })}>
								{row.balance < 0 ? '(' : ''}
								<span>$ </span>
								<span style={{ visibility: 'hidden' }}>{prefix}</span>
								{currencyFormatter(Math.abs(row.balance), true)}
								{row.balance < 0 ? ')' : ''}
							</div>
						</td>
					);
				},
			},
		],
	};

	return (
		<>
			<h2 className="margin-bottom-neg-205">Monthly utilization</h2>
			<Table {...tableProps} fullWidth />
		</>
	);
}
