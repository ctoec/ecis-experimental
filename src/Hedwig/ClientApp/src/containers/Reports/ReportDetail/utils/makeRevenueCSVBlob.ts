import { parse as parseCsv } from 'json2csv';
import { CdcReport, FundingTime } from "../../../../generated";
import { prettyAge, prettyFundingTime, getReportingPeriodWeeks } from "../../../../utils/models";
import { getRevenueDataRows } from "./getRevenueDataRows";

export const makeRevenueCSVBlob = (report: CdcReport) => {
	let revenueData = getRevenueDataRows(report);
	const totalRow = revenueData.find((row) => row.key === 'total');

	const revenueCSVRows: any[] = revenueData.map((row) => ({
		...row,
		title:
			row.key === 'total'
				? 'Total'
				: `${prettyAge(row.ageGroup)} - ${prettyFundingTime(row.fundingTime)}`,
	}));

	revenueCSVRows.push(
		{}, // empty row for final CSV formatting
		{ title: 'Other revenue' },
		{}, // empty row for final CSV formatting
		{ title: 'Care 4 Kids', total: report.c4KRevenue },
		{ title: 'Family Fees', total: report.familyFeesRevenue },
		{}, // empty row for final CSV formatting
		{
			title: 'Total revenue',
			total: (totalRow?.total || 0) + (report.c4KRevenue || 0) + (report.familyFeesRevenue || 0),
		}
	);

	const fields = [
		{
			label: 'CDC Revenue',
			value: 'title',
		},
		{
			label: 'Utilization',
			value: (row: any) =>
				row.count != undefined && row.capacity != undefined ? `${row.count} / ${row.capacity}` : '',
		},
		{
			label: 'Reimbursement rate',
			value: (row: any) => {
				let rateStr = '';
				if (row.ptRate != undefined && row.fundingTime && row.fundingTime !== FundingTime.Full)
					rateStr += `${row.ptRate} (part-time) `;
				if (row.ftRate != undefined && row.fundingTime && row.fundingTime !== FundingTime.Part)
					rateStr += `${row.ftRate} (full-time) `;
				return rateStr;
			},
		},
		{
			label: `Total (${getReportingPeriodWeeks(report.reportingPeriod)} weeks)`,
			value: 'total',
		},
		{
			label: 'Balance',
			value: (row: any) => (row.balance != undefined ? (row.balance as number).toFixed(2) : ''),
		},
	];

	const revenueCsv = parseCsv(revenueCSVRows, { fields });
	return new Blob([revenueCsv], { type: 'text/csv' });
};
