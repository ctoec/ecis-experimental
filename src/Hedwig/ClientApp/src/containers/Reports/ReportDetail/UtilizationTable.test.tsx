import React from 'react';
import { render } from '@testing-library/react';
import { Age, FundingTime, FundingSource, Enrollment, Region, Funding } from '../../../generated';
import UtilizationTable, { calculateRate } from './UtilizationTable';
import emptyGuid from '../../../utils/emptyGuid';
import cartesianProduct from '../../../utils/cartesianProduct';
import { accessibilityTestHelper } from '../../../tests/helpers';
import {
	mockReport,
	mockCompleteEnrollment,
	mockPartTimeEnrollment,
	mockFullTimeInfantEnrollment,
	mockPartTimeInfantEnrollment,
	mockDefaultReport,
} from '../../../tests/data';
import { mockFundingSpaces } from '../../../tests/data/fundingSpace';
import { prettyAge, prettyFundingTime } from '../../../utils/models';

describe('calculateRate', () => {
	it('includes all possible rates', () => {
		cartesianProduct({
			accredited: [true, false],
			titleI: [true, false],
			region: [
				Region.East,
				Region.NorthCentral,
				Region.NorthWest,
				Region.SouthCentral,
				Region.SouthWest,
			],
			ageGroup: [Age.InfantToddler, Age.Preschool, Age.SchoolAge],
			time: [FundingTime.Full, FundingTime.Part],
		}).forEach(combo => {
			const rate = calculateRate(
				combo.accredited,
				combo.titleI,
				combo.region,
				combo.ageGroup,
				combo.time
			);
			expect(rate).toBeGreaterThan(0);
		});
	});
});

const reportWithEnrollments = (enrollments: Enrollment[]) => {
	return { ...mockReport, enrollments };
};

const defaultReport = reportWithEnrollments([mockCompleteEnrollment]);

describe('UtilizationTable', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(<UtilizationTable {...defaultReport} />);
		expect(asFragment()).toMatchSnapshot();
	});

	it('includes a row for each type of funding space', () => {
		const report = mockReport;
		const { container } = render(<UtilizationTable {...report} />);

		// mockReport belongs to mockSingleSiteOrganization, which has all mockFundingSpaces
		mockFundingSpaces.forEach(space => {
			expect(container).toHaveTextContent(
				`${prettyAge(space.ageGroup)} â€“ ${prettyFundingTime(space.time)}`
			);
		});
	});

	it('has the correct utilization ratios', () => {
		// Default report has 2 full time preschool spaces
		// These mocked enrollments are one of each full and part infant/toddler and preschool
		const mockReport = {
			...mockDefaultReport,
			enrollments: [
				mockPartTimeInfantEnrollment,
				mockFullTimeInfantEnrollment,
				mockCompleteEnrollment, // full time, preschool
				mockPartTimeEnrollment, // part time, preschool
			],
		};

		const { getAllByText } = render(<UtilizationTable {...mockReport} />);
		const oneOfZeros = getAllByText(/1\/\d* spaces/);
		expect(oneOfZeros).toHaveLength(4);
	});

	it('does not include enrollments without an age', () => {
		const report = reportWithEnrollments([
			{
				id: 1,
				ageGroup: undefined,
				siteId: 1,
				childId: emptyGuid(),
				fundings: [
					{
						id: 1,
						source: FundingSource.CDC,
						enrollmentId: 1,
						fundingSpace: {
							organizationId: 1,
							capacity: 1,
							time: FundingTime.Split,
							timeSplit: { fullTimeWeeks: 10, partTimeWeeks: 42 },
						},
					} as Funding,
				],
			},
		]);

		const { container } = render(<UtilizationTable {...report} />);

		expect(container).toHaveTextContent('0/10 spaces');
	});

	accessibilityTestHelper(<UtilizationTable {...defaultReport} />);
});
