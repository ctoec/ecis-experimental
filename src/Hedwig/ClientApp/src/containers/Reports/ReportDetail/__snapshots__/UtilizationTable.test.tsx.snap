// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UtilizationTable matches snapshot 1`] = `
<DocumentFragment>
  <h2
    class="margin-bottom-neg-205"
  >
    Monthly utilization
  </h2>
  <table
    class="oec-table oec-table--full-width"
    id="utilization-table"
  >
    <thead>
      <tr>
        <th
          aria-sort="none"
          class="oec-table__column-header"
          role="columnheader"
          scope="col"
        >
          <span
            class="oec-table__column-title"
          />
        </th>
        <th
          aria-sort="none"
          class="oec-table__column-header"
          role="columnheader"
          scope="col"
        >
          <span
            class="oec-table__column-title"
          >
            Utilization
          </span>
        </th>
        <th
          aria-sort="none"
          class="oec-table__column-header text-right"
          role="columnheader"
          scope="col"
        >
          <span
            class="oec-table__column-title"
          >
            Reimbursement rate
          </span>
        </th>
        <th
          aria-sort="none"
          class="oec-table__column-header text-right"
          role="columnheader"
          scope="col"
        >
          <span
            class="oec-table__column-title"
          >
            Total (4 weeks)
          </span>
        </th>
        <th
          aria-sort="none"
          class="oec-table__column-header text-right"
          role="columnheader"
          scope="col"
        >
          <span
            class="oec-table__column-title"
          >
            Balance
          </span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>
          <strong>
            Infant/Toddler
          </strong>
           – full time
        </th>
        <td
          class=" oec-table__cell--red"
        >
          0/10 spaces
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          />
          169.95
          <span>
             × 4 weeks
          </span>
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            00
          </span>
          0.00
        </td>
        <td
          class="oec-table__cell--red text-tabular text-right"
        >
          <div
            class="position-relative one-half-char-right"
          >
            (
            <span>
              $ 
            </span>
            <span
              style="visibility: hidden;"
            >
              0
            </span>
            6,798.00)
          </div>
        </td>
      </tr>
      <tr>
        <th>
          <strong>
            Infant/Toddler
          </strong>
           – part time
        </th>
        <td
          class=" oec-table__cell--red"
        >
          0/10 spaces
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            0
          </span>
          59.50
          <span>
             × 4 weeks
          </span>
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            00
          </span>
          0.00
        </td>
        <td
          class="oec-table__cell--red text-tabular text-right"
        >
          <div
            class="position-relative one-half-char-right"
          >
            (
            <span>
              $ 
            </span>
            <span
              style="visibility: hidden;"
            >
              0
            </span>
            2,380.00)
          </div>
        </td>
      </tr>
      <tr>
        <th>
          <strong>
            Infant/Toddler
          </strong>
           – split time
        </th>
        <td
          class=" oec-table__cell--red"
        >
          0/10 spaces
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            00
          </span>
          0.00
          <span>
             × 4 weeks
          </span>
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            00
          </span>
          0.00
        </td>
        <td
          class="text-tabular text-right"
        >
          <div
            class="position-relative"
          >
            <span>
              $ 
            </span>
            <span
              style="visibility: hidden;"
            >
              0000,
            </span>
            0.00
          </div>
        </td>
      </tr>
      <tr>
        <th>
          <strong>
            Preschool
          </strong>
           – full time
        </th>
        <td
          class=" oec-table__cell--red"
        >
          1/10 spaces
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          />
          165.32
          <span>
             × 4 weeks
          </span>
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          />
          661.28
        </td>
        <td
          class="oec-table__cell--red text-tabular text-right"
        >
          <div
            class="position-relative one-half-char-right"
          >
            (
            <span>
              $ 
            </span>
            <span
              style="visibility: hidden;"
            >
              0
            </span>
            5,951.52)
          </div>
        </td>
      </tr>
      <tr>
        <th>
          <strong>
            Preschool
          </strong>
           – part time
        </th>
        <td
          class=" oec-table__cell--red"
        >
          0/10 spaces
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            0
          </span>
          57.85
          <span>
             × 4 weeks
          </span>
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            00
          </span>
          0.00
        </td>
        <td
          class="oec-table__cell--red text-tabular text-right"
        >
          <div
            class="position-relative one-half-char-right"
          >
            (
            <span>
              $ 
            </span>
            <span
              style="visibility: hidden;"
            >
              0
            </span>
            2,314.00)
          </div>
        </td>
      </tr>
      <tr>
        <th>
          <strong>
            School-age
          </strong>
           – full time
        </th>
        <td
          class=" oec-table__cell--red"
        >
          0/10 spaces
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          />
          113.20
          <span>
             × 4 weeks
          </span>
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            00
          </span>
          0.00
        </td>
        <td
          class="oec-table__cell--red text-tabular text-right"
        >
          <div
            class="position-relative one-half-char-right"
          >
            (
            <span>
              $ 
            </span>
            <span
              style="visibility: hidden;"
            >
              0
            </span>
            4,528.00)
          </div>
        </td>
      </tr>
      <tr>
        <th>
          <strong>
            School-age
          </strong>
           – part time
        </th>
        <td
          class=" oec-table__cell--red"
        >
          0/10 spaces
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            0
          </span>
          54.80
          <span>
             × 4 weeks
          </span>
        </td>
        <td
          class="text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          >
            00
          </span>
          0.00
        </td>
        <td
          class="oec-table__cell--red text-tabular text-right"
        >
          <div
            class="position-relative one-half-char-right"
          >
            (
            <span>
              $ 
            </span>
            <span
              style="visibility: hidden;"
            >
              0
            </span>
            2,192.00)
          </div>
        </td>
      </tr>
      <tr>
        <th>
          <strong>
            Total
          </strong>
        </th>
        <td
          class="oec-table__cell--strong oec-table__cell--red"
        >
          1/70 spaces
        </td>
        <td
          class="text-tabular text-right"
        />
        <td
          class="oec-table__cell--strong text-tabular text-right"
        >
          <span>
            $ 
          </span>
          <span
            style="visibility: hidden;"
          />
          661.28
        </td>
        <td
          class="oec-table__cell--strong oec-table__cell--red text-tabular text-right"
        >
          <div
            class="position-relative one-half-char-right"
          >
            (
            <span>
              $ 
            </span>
            <span
              style="visibility: hidden;"
            />
            24,163.52)
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
